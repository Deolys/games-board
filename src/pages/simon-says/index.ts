export { SimonSaysGame } from "./simon-says";