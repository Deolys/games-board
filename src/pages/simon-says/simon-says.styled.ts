import styled from '@emotion/styled';
import { Button, Paper } from '@mui/material';

export const ColorButton = styled(Button)({
  margin: 1,
  width: 100,
  height: 100,
});

export const GameBoard = styled(Paper)({
  display: 'grid',
  gridTemplateColumns: 'repeat(2, 100px)',
  gap: 2,
  alignItems: 'center',
  justifyContent: 'center',
  padding: 30,
});