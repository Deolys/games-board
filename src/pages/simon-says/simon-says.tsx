import { useState, useEffect, useCallback } from 'react';
import { Box, Button, Snackbar, Alert } from '@mui/material';
import { Header } from '../../components/header';
import { ColorButton, GameBoard } from './simon-says.styled';

const colors = ['green', 'red', 'yellow', 'blue'];

export function SimonSaysGame() {
  const [sequence, setSequence] = useState<string[]>([colors[Math.floor(Math.random() * colors.length)]]);
  const [userSequence, setUserSequence] = useState<string[]>([]);
  const [activeColor, setActiveColor] = useState<string>('');
  const [open, setOpen] = useState<boolean>(false);
  const [message, setMessage] = useState<string>('');

  const flashSequence = (sequence: string[]): void => {
    sequence.forEach((color, index) => {
      setTimeout(() => {
        setActiveColor(color);
        setTimeout(() => setActiveColor(''), 500);
      }, (index + 1) * 1000);
    });
  };

  const nextLevel = useCallback(() => {
    const nextColor = colors[Math.floor(Math.random() * colors.length)];
    setSequence([...sequence, nextColor]);
    setUserSequence([]);
    flashSequence([...sequence, nextColor]);
  }, [sequence]);

  const checkSequence = useCallback(() => {
    if (userSequence.length === 0) return;

    const isCorrect: boolean = userSequence.every((val, index) => val === sequence[index]);
    if (userSequence.length === sequence.length && isCorrect) {
      setMessage('Вы правильно повторили последовательность!');
      setOpen(true);
      nextLevel();
    } 
    else if (!isCorrect) {
      setMessage('Неправильно, попробуйте еще раз!');
      setOpen(true);
      setUserSequence([]);
    }
  }, [sequence, userSequence, nextLevel]);

  useEffect(() => {
    if (sequence.length === 0) {
      nextLevel()
      return;
    }
    checkSequence();
  }, [checkSequence, sequence, nextLevel]);

  const handleClose = (_event?: React.SyntheticEvent | Event, reason?: string) => {
    if (reason === 'clickaway') {
      return;
    }
    setOpen(false);
  };

  const handleColorClick = (color: string): void => {
    setUserSequence([...userSequence, color]);
    setActiveColor(color);
    setTimeout(() => setActiveColor(''), 500);
  };

  const startGame = (): void => {
    setSequence([]);
    setUserSequence([]);
  };

  return (
   <>
   <Header />
   <Box component="section" sx={{ display: 'flex', justifyContent: 'center', flexDirection: "column", height: '88vh' }}>
      <GameBoard elevation={3}>
        {colors.map(color => (
          <ColorButton
            data-testid={color}
            key={color}
            style={{ backgroundColor: activeColor === color ? color : 'grey' }}
            onClick={() => handleColorClick(color)}
          />
        ))}
      </GameBoard>
      <Box sx={{ mt: 2, textAlign: 'center'}}>
        <Button variant="contained" color="primary" onClick={startGame}>
          Начать игру
        </Button>
      </Box>
      <Snackbar open={open} autoHideDuration={4000} onClose={handleClose} anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}>
        <Alert onClose={handleClose} severity="info" sx={{ width: '100%' }}>
          {message}
        </Alert>
      </Snackbar>
    </Box>
   </>
  );
}

export default SimonSaysGame;
