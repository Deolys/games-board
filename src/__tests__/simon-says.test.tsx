import { fireEvent, render, screen } from "@testing-library/react";
import { SimonSaysGame } from "../pages/simon-says";

describe('SimonSaysGame', () => {
  it('SimonSaysGame mounts properly', () => {
    const { container } = render(<SimonSaysGame />);
    expect(container).toMatchSnapshot();
  })

  it('Button works', () => {
    render(<SimonSaysGame />);
    const colorButton = screen.getByTestId('red');
    fireEvent.click(colorButton);
  })
});