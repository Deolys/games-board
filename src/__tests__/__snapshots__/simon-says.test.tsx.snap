// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimonSaysGame SimonSaysGame mounts properly 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1413kmy-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-u4fbkj-MuiToolbar-root"
    >
      <a
        class="css-3qnz2e"
        href="/"
      >
        <img
          alt="Games board"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-1waifce-MuiTypography-root"
        >
          Games board
        </p>
      </a>
      <div
        class="MuiStack-root css-1dzrg60-MuiStack-root"
      >
        <a
          class="MuiTypography-root MuiTypography-body1 css-1v5m6us-MuiTypography-root"
          href="/"
        >
          HOME
        </a>
        <button
          aria-haspopup="true"
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-17ovnxk-MuiButtonBase-root-MuiButton-root"
          id="composition-button"
          tabindex="0"
          type="button"
        >
          Games
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
  <section
    class="MuiBox-root css-cmz0bm"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-1f7efoc-MuiPaper-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-tfhrva-MuiButtonBase-root-MuiButton-root"
        data-testid="green"
        style="background-color: grey;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-tfhrva-MuiButtonBase-root-MuiButton-root"
        data-testid="red"
        style="background-color: grey;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-tfhrva-MuiButtonBase-root-MuiButton-root"
        data-testid="yellow"
        style="background-color: grey;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-tfhrva-MuiButtonBase-root-MuiButton-root"
        data-testid="blue"
        style="background-color: grey;"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiBox-root css-qwujcl"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Начать игру
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </section>
</div>
`;
